File: pkg/terminalservice/protocol.go
package terminalservice

import "encoding/json"

const ProtocolVersion = 2

type ControlRequest struct {
	ProtocolVersion int             `json:"protocolVersion"`
	Method          string          `json:"method"`
	Params          json.RawMessage `json:"params,omitempty"`
}

type ControlResponse struct {
	OK     bool   `json:"ok"`
	Result any    `json:"result,omitempty"`
	Error  string `json:"error,omitempty"`
}

type CreateRequest struct {
	SessionID string `json:"sessionId"`
	Cwd       string `json:"cwd"`
}

type CreateResponse struct {
	SessionID string `json:"sessionId"`
	Existing  bool   `json:"existing"`
}

type SendRequest struct {
	SessionID string `json:"sessionId"`
	Data      string `json:"data"`
}

type ResizeRequest struct {
	SessionID string `json:"sessionId"`
	Cols      int    `json:"cols"`
	Rows      int    `json:"rows"`
}

type StopRequest struct {
	SessionID string `json:"sessionId"`
}

type InspectRequest struct {
	SessionID string `json:"sessionId"`
}

type ShutdownRequest struct {
	Source     string `json:"source,omitempty"`
	Reason     string `json:"reason,omitempty"`
	PID        int    `json:"pid,omitempty"`
	Executable string `json:"executable,omitempty"`
}

type SessionInfo struct {
	SessionID  string `json:"sessionId"`
	Cwd        string `json:"cwd"`
	StartedAt  string `json:"startedAt"`
	LastActive string `json:"lastActive"`
	Running    bool   `json:"running"`
}

type InspectResponse struct {
	SessionID  string `json:"sessionId"`
	Cwd        string `json:"cwd"`
	StartedAt  string `json:"startedAt"`
	LastActive string `json:"lastActive"`
	Running    bool   `json:"running"`
}

type ListResponse struct {
	Sessions []SessionInfo `json:"sessions"`
}

type AttachRequest struct {
	ProtocolVersion int    `json:"protocolVersion"`
	Type            string `json:"type"`
	SessionID       string `json:"sessionId"`
	StreamID        string `json:"streamId,omitempty"`
	Token           string `json:"token,omitempty"`
}

type WebsocketControlRequest struct {
	ProtocolVersion int    `json:"protocolVersion,omitempty"`
	Type            string `json:"type"`
	Data            string `json:"data,omitempty"`
	Cols            int    `json:"cols,omitempty"`
	Rows            int    `json:"rows,omitempty"`
}

type InfoResponse struct {
	Executable     string `json:"executable"`
	BinaryHash     string `json:"binaryHash"`
	WebSocketURL   string `json:"webSocketUrl,omitempty"`
	WebSocketToken string `json:"webSocketToken,omitempty"`
}

type StreamMessage struct {
	Type       string `json:"type"`
	SessionID  string `json:"sessionId,omitempty"`
	StreamID   string `json:"streamId,omitempty"`
	DataB64    string `json:"dataB64,omitempty"`
	Len        int    `json:"len,omitempty"`
	NextOffset int64  `json:"nextOffset,omitempty"`
	Error      string `json:"error,omitempty"`
}

File: OPUS-tauri-app/src-tauri/src/sessiond/protocol.rs
use serde::{Deserialize, Serialize};

pub const PROTOCOL_VERSION: u32 = 2;

// ---------- Control request/response ----------

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ControlRequest {
    pub protocol_version: u32,
    pub method: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub params: Option<serde_json::Value>,
}

#[derive(Debug, Deserialize)]
pub struct ControlResponse {
    pub ok: bool,
    #[serde(default)]
    pub result: serde_json::Value,
    #[serde(default)]
    pub error: Option<String>,
}

// ---------- Create ----------

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateRequest {
    pub session_id: String,
    pub cwd: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateResponse {
    pub session_id: String,
    #[serde(default)]
    pub existing: bool,
}

// ---------- Attach ----------

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AttachRequest {
    pub protocol_version: u32,
    #[serde(rename = "type")]
    pub msg_type: String,
    pub session_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub stream_id: Option<String>,
    pub since: i64,
    pub with_buffer: bool,
}

// ---------- Stream message (received from sessiond) ----------

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StreamMessage {
    #[serde(rename = "type")]
    pub msg_type: String,
    #[serde(default)]
    pub session_id: String,
    #[serde(default)]
    pub stream_id: String,
    #[serde(default)]
    pub data: Option<String>,
    #[serde(default)]
    pub len: Option<usize>,
    #[serde(default)]
    pub next_offset: Option<i64>,
    #[serde(default)]
    pub truncated: bool,
    #[serde(default)]
    pub source: Option<String>,
    #[serde(default)]
    pub snapshot_source: Option<String>,
    #[serde(default)]
    pub backlog_source: Option<String>,
    #[serde(default)]
    pub backlog_truncated: bool,
    #[serde(default)]
    pub alt_screen: bool,
    #[serde(default)]
    pub mouse_mask: u8,
    #[serde(default)]
    pub mouse: bool,
    #[serde(default)]
    pub mouse_sgr: bool,
    #[serde(default)]
    pub mouse_encoding: Option<String>,
    #[serde(default)]
    pub safe_to_replay: bool,
    #[serde(default)]
    pub initial_credit: Option<i64>,
    #[serde(default)]
    pub error: Option<String>,
}

// ---------- Send / Resize / Stop / ACK ----------

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SendRequest {
    pub session_id: String,
    pub data: String,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ResizeRequest {
    pub session_id: String,
    pub cols: u32,
    pub rows: u32,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StopRequest {
    pub session_id: String,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AckRequest {
    pub session_id: String,
    pub stream_id: String,
    pub bytes: i64,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BootstrapRequest {
    pub session_id: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BootstrapResponse {
    pub session_id: String,
    #[serde(default)]
    pub snapshot: Option<String>,
    #[serde(default)]
    pub backlog: Option<String>,
    #[serde(default)]
    pub next_offset: Option<i64>,
    #[serde(default)]
    pub backlog_truncated: bool,
    #[serde(default)]
    pub alt_screen: bool,
    #[serde(default)]
    pub mouse: bool,
    #[serde(default)]
    pub mouse_sgr: bool,
    #[serde(default)]
    pub mouse_encoding: Option<String>,
    #[serde(default)]
    pub safe_to_replay: bool,
    #[serde(default)]
    pub initial_credit: Option<i64>,
}
